<!DOCTYPE html>
<html data-color-mode="light" data-dark-theme="dark_colorblind" data-light-theme="light" lang="zh-CN">
<head>
    <meta content="text/html; charset=utf-8" http-equiv="content-type" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <link href='https://mirrors.sustech.edu.cn/cdnjs/ajax/libs/Primer/21.0.7/primer.css' rel='stylesheet' />
    
    <link rel="icon" href="https://blog.00002000.xyz/logo.svg"><script>
        let theme = localStorage.getItem("meek_theme") || "light";
        document.documentElement.setAttribute("data-color-mode", theme);
    </script>
<meta name="description" content="| 软件名称 | 简介 | 教程 | 平台(PC) | tags |
| -------- | ---- | ---- | -------- | ---- |
| trimAl    | 大规模系统发育分析的自动比对修剪工具     | #40  | 跨平台(AUR)   | ![GitHub top language](https://img.shields.io/github/languages/top/inab/trimal) ![GitHub last commit](https://img.shields.io/github/last-commit/inab/trimal) ![GitHub issues](https://img.shields.io/github/issues/inab/trimal) ![GitHub stars](https://img.shields.io/github/stars/inab/trimal?style=flat) ![GitHub commit activity](https://img.shields.io/github/commit-activity/t/inab/trimal) ![GitHub contributors](https://img.shields.io/github/contributors/inab/trimal) ![GitHub Created At](https://img.shields.io/github/created-at/inab/trimal) ![GitHub forks](https://img.shields.io/github/forks/inab/trimal?style=flat) ![GitHub pull requests](https://img.shields.io/github/issues-pr/inab/trimal) ![GitHub license](https://img.shields.io/github/license/inab/trimal) ![GitHub release](https://img.shields.io/github/v/release/inab/trimal) ![GitHub language count](https://img.shields.io/github/languages/count/inab/trimal)   |

- [trimAl — trimAl](https://trimal.readthedocs.io/en/latest/)
- [inab/trimal: A tool for automated alignment trimming in large-scale phylogenetic analyses. Development version: 2.0](https://github.com/inab/trimal)
- [AUR (en) - trimal](https://aur.archlinux.org/packages/trimal)：![AUR 最后修改](https://img.shields.io/aur/last-modified/trimal) ![AUR 版本](https://img.shields.io/aur/version/trimal) ![AUR 许可证](https://img.shields.io/aur/license/trimal) ![AUR 维护者](https://img.shields.io/aur/maintainer/trimal) ![AUR 投票](https://img.shields.io/aur/votes/trimal) ![AUR 依赖](https://img.shields.io/badge/dynamic/json?url=https%3A%2F%2Faur.archlinux.org%2Frpc%2Fv5%2Finfo%2Ftrimal&query=%24..Depends&label=Depends) ![AUR 编译依赖](https://img.shields.io/badge/dynamic/json?url=https%3A%2F%2Faur.archlinux.org%2Frpc%2Fv5%2Finfo%2Ftrimal&query=%24..MakeDepends&label=MakeDepends)

## 简介

trimAl，全称为“trimAl”（Trimming Alignment with Logical Constraints），是一款生物信息学软件，专门用于优化多序列比对（Multiple Sequence Alignments，MSA）。">
<meta property="og:title" content="trimAl：大规模系统发育分析的自动比对修剪工具">
<meta property="og:description" content="| 软件名称 | 简介 | 教程 | 平台(PC) | tags |
| -------- | ---- | ---- | -------- | ---- |
| trimAl    | 大规模系统发育分析的自动比对修剪工具     | #40  | 跨平台(AUR)   | ![GitHub top language](https://img.shields.io/github/languages/top/inab/trimal) ![GitHub last commit](https://img.shields.io/github/last-commit/inab/trimal) ![GitHub issues](https://img.shields.io/github/issues/inab/trimal) ![GitHub stars](https://img.shields.io/github/stars/inab/trimal?style=flat) ![GitHub commit activity](https://img.shields.io/github/commit-activity/t/inab/trimal) ![GitHub contributors](https://img.shields.io/github/contributors/inab/trimal) ![GitHub Created At](https://img.shields.io/github/created-at/inab/trimal) ![GitHub forks](https://img.shields.io/github/forks/inab/trimal?style=flat) ![GitHub pull requests](https://img.shields.io/github/issues-pr/inab/trimal) ![GitHub license](https://img.shields.io/github/license/inab/trimal) ![GitHub release](https://img.shields.io/github/v/release/inab/trimal) ![GitHub language count](https://img.shields.io/github/languages/count/inab/trimal)   |

- [trimAl — trimAl](https://trimal.readthedocs.io/en/latest/)
- [inab/trimal: A tool for automated alignment trimming in large-scale phylogenetic analyses. Development version: 2.0](https://github.com/inab/trimal)
- [AUR (en) - trimal](https://aur.archlinux.org/packages/trimal)：![AUR 最后修改](https://img.shields.io/aur/last-modified/trimal) ![AUR 版本](https://img.shields.io/aur/version/trimal) ![AUR 许可证](https://img.shields.io/aur/license/trimal) ![AUR 维护者](https://img.shields.io/aur/maintainer/trimal) ![AUR 投票](https://img.shields.io/aur/votes/trimal) ![AUR 依赖](https://img.shields.io/badge/dynamic/json?url=https%3A%2F%2Faur.archlinux.org%2Frpc%2Fv5%2Finfo%2Ftrimal&query=%24..Depends&label=Depends) ![AUR 编译依赖](https://img.shields.io/badge/dynamic/json?url=https%3A%2F%2Faur.archlinux.org%2Frpc%2Fv5%2Finfo%2Ftrimal&query=%24..MakeDepends&label=MakeDepends)

## 简介

trimAl，全称为“trimAl”（Trimming Alignment with Logical Constraints），是一款生物信息学软件，专门用于优化多序列比对（Multiple Sequence Alignments，MSA）。">
<meta property="og:type" content="article">
<meta property="og:url" content="https://blog.00002000.xyz/post/40.html">
<meta property="og:image" content="https://blog.00002000.xyz/logo.svg">
<title>trimAl：大规模系统发育分析的自动比对修剪工具</title>
<link href="//unpkg.com/@wooorm/starry-night@2.1.1/style/both.css" rel="stylesheet" />


</head>
<style>
body{box-sizing: border-box;min-width: 200px;max-width: 900px;margin: 20px auto;padding: 45px;font-size: 16px;font-family: sans-serif;line-height: 1.25;}
#header{display:flex;padding-bottom:8px;border-bottom: 1px solid var(--borderColor-muted, var(--color-border-muted));margin-bottom: 16px;}
#footer {margin-top:64px; text-align: center;font-size: small;}

</style>

<style>
.postTitle{margin: auto 0;font-size:40px;font-weight:bold;}
.title-right{display:flex;margin:auto 0 0 auto;}
.title-right .circle{padding: 14px 16px;margin-right:8px;}
#postBody{border-bottom: 1px solid var(--color-border-default);padding-bottom:36px;}
#postBody hr{height:2px;}
#cmButton{height:48px;margin-top:48px;}
#comments{margin-top:64px;}
.g-emoji{font-size:24px;}
@media (max-width: 600px) {
    body {padding: 8px;}
    .postTitle{font-size:24px;}
}
.copy-feedback {
    display: none;
    position: absolute;
    top: 10px;
    right: 50px;
    color: var(--color-fg-on-emphasis);
    background-color: var(--color-fg-muted);
    border-radius: 3px;
    padding: 5px 8px;
    font-size: 12px;
}
</style>




<body>
    <div id="header">
<h1 class="postTitle">trimAl：大规模系统发育分析的自动比对修剪工具</h1>
<div class="title-right">
    <a href="https://blog.00002000.xyz" id="buttonHome" class="btn btn-invisible circle" title="首页">
        <svg class="octicon" width="16" height="16">
            <path id="pathHome" fill-rule="evenodd"></path>
        </svg>
    </a>
    
    <a href="https://github.com/seeu100/blog/issues/40" target="_blank" class="btn btn-invisible circle" title="Issue">
        <svg class="octicon" width="16" height="16">
            <path id="pathIssue" fill-rule="evenodd"></path>
        </svg>
    </a>
    

    <a class="btn btn-invisible circle" onclick="modeSwitch();" title="切换主题">
        <svg class="octicon" width="16" height="16" >
            <path id="themeSwitch" fill-rule="evenodd"></path>
        </svg>
    </a>

</div>
</div>
    <div id="content">
<div class="markdown-body" id="postBody"><markdown-accessiblity-table><table role="table">
<thead>
<tr>
<th>软件名称</th>
<th>简介</th>
<th>教程</th>
<th>平台(PC)</th>
<th>tags</th>
</tr>
</thead>
<tbody>
<tr>
<td>trimAl</td>
<td>大规模系统发育分析的自动比对修剪工具</td>
<td>#40</td>
<td>跨平台(AUR)</td>
<td><a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/9b1008becdfad15ccf69eaee47acde251b587c169c881eb464704525b437a37e/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6c616e6775616765732f746f702f696e61622f7472696d616c"><img src="https://camo.githubusercontent.com/9b1008becdfad15ccf69eaee47acde251b587c169c881eb464704525b437a37e/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6c616e6775616765732f746f702f696e61622f7472696d616c" alt="GitHub top language" data-canonical-src="https://img.shields.io/github/languages/top/inab/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/ed78a4fc1a7e45654e806b377362cd76ac16d1d71f77a8a9b35be1f9ad860c39/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6c6173742d636f6d6d69742f696e61622f7472696d616c"><img src="https://camo.githubusercontent.com/ed78a4fc1a7e45654e806b377362cd76ac16d1d71f77a8a9b35be1f9ad860c39/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6c6173742d636f6d6d69742f696e61622f7472696d616c" alt="GitHub last commit" data-canonical-src="https://img.shields.io/github/last-commit/inab/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/b243bb61f3148d6c12c02ee317d2e778823ac8bb4f584832aeb8b1d007e86b51/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6973737565732f696e61622f7472696d616c"><img src="https://camo.githubusercontent.com/b243bb61f3148d6c12c02ee317d2e778823ac8bb4f584832aeb8b1d007e86b51/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6973737565732f696e61622f7472696d616c" alt="GitHub issues" data-canonical-src="https://img.shields.io/github/issues/inab/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/fc4f898bc3fb4c7cb5dd77665921dbe0d2405a665895144e2e04aa421cbd686c/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f73746172732f696e61622f7472696d616c3f7374796c653d666c6174"><img src="https://camo.githubusercontent.com/fc4f898bc3fb4c7cb5dd77665921dbe0d2405a665895144e2e04aa421cbd686c/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f73746172732f696e61622f7472696d616c3f7374796c653d666c6174" alt="GitHub stars" data-canonical-src="https://img.shields.io/github/stars/inab/trimal?style=flat" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/8c9f59dc690ff14c3c81a340576121e2b9f9b66fcba5dfbb7fd33753d874e5ad/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f636f6d6d69742d61637469766974792f742f696e61622f7472696d616c"><img src="https://camo.githubusercontent.com/8c9f59dc690ff14c3c81a340576121e2b9f9b66fcba5dfbb7fd33753d874e5ad/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f636f6d6d69742d61637469766974792f742f696e61622f7472696d616c" alt="GitHub commit activity" data-canonical-src="https://img.shields.io/github/commit-activity/t/inab/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/13899ada447c25035840d66f8b8e3d29f715352c8c103b367e244559fa434a67/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f636f6e7472696275746f72732f696e61622f7472696d616c"><img src="https://camo.githubusercontent.com/13899ada447c25035840d66f8b8e3d29f715352c8c103b367e244559fa434a67/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f636f6e7472696275746f72732f696e61622f7472696d616c" alt="GitHub contributors" data-canonical-src="https://img.shields.io/github/contributors/inab/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/3ab9a36702f314e1aee63699874dba39a332eeefad177e0900a3a3302d795923/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f637265617465642d61742f696e61622f7472696d616c"><img src="https://camo.githubusercontent.com/3ab9a36702f314e1aee63699874dba39a332eeefad177e0900a3a3302d795923/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f637265617465642d61742f696e61622f7472696d616c" alt="GitHub Created At" data-canonical-src="https://img.shields.io/github/created-at/inab/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/a11e034aa7c0271b9ccb2008e642b70fbd1fe53c3e777f3cddf0755edab09839/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f666f726b732f696e61622f7472696d616c3f7374796c653d666c6174"><img src="https://camo.githubusercontent.com/a11e034aa7c0271b9ccb2008e642b70fbd1fe53c3e777f3cddf0755edab09839/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f666f726b732f696e61622f7472696d616c3f7374796c653d666c6174" alt="GitHub forks" data-canonical-src="https://img.shields.io/github/forks/inab/trimal?style=flat" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/b93f91cd77f3e301cedeb6085fa2c5f0355869fe00c479503401abd527126cc6/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6973737565732d70722f696e61622f7472696d616c"><img src="https://camo.githubusercontent.com/b93f91cd77f3e301cedeb6085fa2c5f0355869fe00c479503401abd527126cc6/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6973737565732d70722f696e61622f7472696d616c" alt="GitHub pull requests" data-canonical-src="https://img.shields.io/github/issues-pr/inab/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/6db2816ffb3baa6a9be639a7d2fdd44de3c34b94cee1153d28bf7ec566fd8a9f/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6c6963656e73652f696e61622f7472696d616c"><img src="https://camo.githubusercontent.com/6db2816ffb3baa6a9be639a7d2fdd44de3c34b94cee1153d28bf7ec566fd8a9f/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6c6963656e73652f696e61622f7472696d616c" alt="GitHub license" data-canonical-src="https://img.shields.io/github/license/inab/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/093a4f99796a8fcbbf6ec87e873370e209f08a907e4e4cec9996113067312624/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f762f72656c656173652f696e61622f7472696d616c"><img src="https://camo.githubusercontent.com/093a4f99796a8fcbbf6ec87e873370e209f08a907e4e4cec9996113067312624/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f762f72656c656173652f696e61622f7472696d616c" alt="GitHub release" data-canonical-src="https://img.shields.io/github/v/release/inab/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/409bdc34104b4473a3dd2969b263771627a8761f355152496cf534db061b8842/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6c616e6775616765732f636f756e742f696e61622f7472696d616c"><img src="https://camo.githubusercontent.com/409bdc34104b4473a3dd2969b263771627a8761f355152496cf534db061b8842/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6c616e6775616765732f636f756e742f696e61622f7472696d616c" alt="GitHub language count" data-canonical-src="https://img.shields.io/github/languages/count/inab/trimal" style="max-width: 100%;"></a></td>
</tr>
</tbody>
</table></markdown-accessiblity-table>
<ul>
<li><a href="https://trimal.readthedocs.io/en/latest/" rel="nofollow">trimAl — trimAl</a></li>
<li><a href="https://github.com/inab/trimal">inab/trimal: A tool for automated alignment trimming in large-scale phylogenetic analyses. Development version: 2.0</a></li>
<li><a href="https://aur.archlinux.org/packages/trimal" rel="nofollow">AUR (en) - trimal</a>：<a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/4519a10fd21e64335db39f7396d49b84bb0405dde9b542a212ea6cebb1646f76/68747470733a2f2f696d672e736869656c64732e696f2f6175722f6c6173742d6d6f6469666965642f7472696d616c"><img src="https://camo.githubusercontent.com/4519a10fd21e64335db39f7396d49b84bb0405dde9b542a212ea6cebb1646f76/68747470733a2f2f696d672e736869656c64732e696f2f6175722f6c6173742d6d6f6469666965642f7472696d616c" alt="AUR 最后修改" data-canonical-src="https://img.shields.io/aur/last-modified/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/e8aebfdc98edad3b624b9059a54e336ada759f78eef8b176fd6b4dfb7c018f6d/68747470733a2f2f696d672e736869656c64732e696f2f6175722f76657273696f6e2f7472696d616c"><img src="https://camo.githubusercontent.com/e8aebfdc98edad3b624b9059a54e336ada759f78eef8b176fd6b4dfb7c018f6d/68747470733a2f2f696d672e736869656c64732e696f2f6175722f76657273696f6e2f7472696d616c" alt="AUR 版本" data-canonical-src="https://img.shields.io/aur/version/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/c595c4e60251dbf25f2bd4bd2fd3bd6efd1d8587726ccfa78bd2cb0d867a2262/68747470733a2f2f696d672e736869656c64732e696f2f6175722f6c6963656e73652f7472696d616c"><img src="https://camo.githubusercontent.com/c595c4e60251dbf25f2bd4bd2fd3bd6efd1d8587726ccfa78bd2cb0d867a2262/68747470733a2f2f696d672e736869656c64732e696f2f6175722f6c6963656e73652f7472696d616c" alt="AUR 许可证" data-canonical-src="https://img.shields.io/aur/license/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/3f51dbaf5a1a92087b81a2fa2fb8edc231bcae5b14efaf726d648964415c8180/68747470733a2f2f696d672e736869656c64732e696f2f6175722f6d61696e7461696e65722f7472696d616c"><img src="https://camo.githubusercontent.com/3f51dbaf5a1a92087b81a2fa2fb8edc231bcae5b14efaf726d648964415c8180/68747470733a2f2f696d672e736869656c64732e696f2f6175722f6d61696e7461696e65722f7472696d616c" alt="AUR 维护者" data-canonical-src="https://img.shields.io/aur/maintainer/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/ad11e13007df08dfdc240be284aabb3179620e11aa28dc93bcf6d8d0e873e3ae/68747470733a2f2f696d672e736869656c64732e696f2f6175722f766f7465732f7472696d616c"><img src="https://camo.githubusercontent.com/ad11e13007df08dfdc240be284aabb3179620e11aa28dc93bcf6d8d0e873e3ae/68747470733a2f2f696d672e736869656c64732e696f2f6175722f766f7465732f7472696d616c" alt="AUR 投票" data-canonical-src="https://img.shields.io/aur/votes/trimal" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/780e38dfe0368d207fe279bd05560dd7ed05118e35f727c53ca45b86ab12586c/68747470733a2f2f696d672e736869656c64732e696f2f62616467652f64796e616d69632f6a736f6e3f75726c3d68747470732533412532462532466175722e617263686c696e75782e6f72672532467270632532467635253246696e666f2532467472696d616c2671756572793d2532342e2e446570656e6473266c6162656c3d446570656e6473"><img src="https://camo.githubusercontent.com/780e38dfe0368d207fe279bd05560dd7ed05118e35f727c53ca45b86ab12586c/68747470733a2f2f696d672e736869656c64732e696f2f62616467652f64796e616d69632f6a736f6e3f75726c3d68747470732533412532462532466175722e617263686c696e75782e6f72672532467270632532467635253246696e666f2532467472696d616c2671756572793d2532342e2e446570656e6473266c6162656c3d446570656e6473" alt="AUR 依赖" data-canonical-src="https://img.shields.io/badge/dynamic/json?url=https%3A%2F%2Faur.archlinux.org%2Frpc%2Fv5%2Finfo%2Ftrimal&amp;query=%24..Depends&amp;label=Depends" style="max-width: 100%;"></a> <a target="_blank" rel="noopener noreferrer nofollow" href="https://camo.githubusercontent.com/8c7d5a2ff1058f59207f5410f2bb1b6f796b4c1c298d5e54d08745e584d1b98f/68747470733a2f2f696d672e736869656c64732e696f2f62616467652f64796e616d69632f6a736f6e3f75726c3d68747470732533412532462532466175722e617263686c696e75782e6f72672532467270632532467635253246696e666f2532467472696d616c2671756572793d2532342e2e4d616b65446570656e6473266c6162656c3d4d616b65446570656e6473"><img src="https://camo.githubusercontent.com/8c7d5a2ff1058f59207f5410f2bb1b6f796b4c1c298d5e54d08745e584d1b98f/68747470733a2f2f696d672e736869656c64732e696f2f62616467652f64796e616d69632f6a736f6e3f75726c3d68747470732533412532462532466175722e617263686c696e75782e6f72672532467270632532467635253246696e666f2532467472696d616c2671756572793d2532342e2e4d616b65446570656e6473266c6162656c3d4d616b65446570656e6473" alt="AUR 编译依赖" data-canonical-src="https://img.shields.io/badge/dynamic/json?url=https%3A%2F%2Faur.archlinux.org%2Frpc%2Fv5%2Finfo%2Ftrimal&amp;query=%24..MakeDepends&amp;label=MakeDepends" style="max-width: 100%;"></a></li>
</ul>
<h2>简介</h2>
<p>trimAl，全称为“trimAl”（Trimming Alignment with Logical Constraints），是一款生物信息学软件，专门用于优化多序列比对（Multiple Sequence Alignments，MSA）。它通过智能化筛选和修剪的方法，移除MSA中的低质量区域，包括不可靠的对齐位置和高度变异的序列片段，从而提高系统发育分析和进化分析的质量和可靠性。</p>
<p>trimAl采用了一系列算法来识别和排除不理想的对齐区域，这些区域可能由于噪音信号（如随机匹配、插入删除事件引起的偏差等）而降低了对齐的整体质量。它提供了多种策略和参数，允许用户根据需求选择合适的修剪模式，例如自动模式（automated1、automated2、automated3）以及其他手动设定阈值的方式，来保留高度保守的序列区块。</p>
<p>通过trimAl处理后的多序列比对，可以减少潜在的误差来源，有利于后续的分子进化分析，例如构建更准确的系统发育树、计算进化距离和检测选择压力等。由于trimAl在优化对齐方面的高效性与实用性，它已成为生物学研究中常用的预处理工具，尤其是在处理核苷酸或氨基酸序列数据时不可或缺。</p>
<pre class="notranslate"><code class="notranslate">❯ paru trimal
1 aur/trimal 1.4.1-1 [+0 ~0.00]
    A tool for automated alignment trimming in large-scale phylogenetic analyses
2 aur/python-pytrimal 0.7.0-1 [+0 ~0.00]
    Cython bindings and Python interface to trimAl, a tool for automated alignment trimming.
</code></pre>
<h2>使用方法</h2>
<p>trimAl的使用方法通常涉及命令行界面，通过指定不同的参数来达到修剪多序列比对的目的。以下是一些基本的使用方法和命令参数解释：</p>
<h3>基本使用格式</h3>
<div class="highlight highlight-source-shell"><pre class="notranslate">trimal -in input_alignment.fasta -out trimmed_alignment.fasta [options]</pre></div>
<ul>
<li><code class="notranslate">-in</code> 参数：指定待修剪的输入多序列比对文件，通常是FASTA格式。</li>
<li><code class="notranslate">-out</code> 参数：指定修剪后输出的新比对文件的名称。</li>
</ul>
<h3>主要参数选项</h3>
<ol>
<li><code class="notranslate">-gt</code> 或 <code class="notranslate">--gappyout</code>: 移除所有包含超过指定百分比（例如 <code class="notranslate">-gt 0.5</code> 表示超过50%）-gap的列。</li>
<li><code class="notranslate">-consistency</code>：根据一致性得分修剪对齐，保留一致性高于指定阈值的列。</li>
<li><code class="notranslate">-strict</code>：严格模式，移除任何包含gap的列。</li>
<li><code class="notranslate">- automated1</code>：自动模式1，trimAl内部设定的一套规则，用于去除不保守的序列区域。</li>
<li><code class="notranslate">-threshold</code>：基于相似度阈值，移除低于指定相似度的列。</li>
<li><code class="notranslate">-fasta</code>：输出格式为FASTA，默认情况下也是这样。</li>
<li><code class="notranslate">-no-gaps</code>：去除所有包含gap的序列。</li>
<li><code class="notranslate">-cluster</code>：基于距离进行分群，然后仅保留每一群中最保守的部分。</li>
<li><code class="notranslate">-discard</code>：丢弃指定百分比最低相似度的序列。</li>
<li><code class="notranslate">-trim</code>：根据指定的修剪模式进行操作，例如 <code class="notranslate">-trim gappyout</code>。</li>
</ol>
<p>在使用trimAl时，用户可以根据自己的具体需求选择相应的参数组合，以期得到最适合进一步分析的优化比对结果。具体的参数含义和用法可以通过查阅trimAl的帮助文档或参考软件作者提供的指南获取详细信息。由于trimAl的参数众多，以上只是列举了一些常见参数，实际使用时可能还有更多高级选项和组合。</p>
<h2>文献阅读</h2>
<p><a href="https://academic.oup.com/bioinformatics/article/25/15/1972/213148?login=true" rel="nofollow">trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses | Bioinformatics | Oxford Academic</a><br>
<a href="https://doi.org/10.1093/bioinformatics/btp348" rel="nofollow">10.1093/bioinformatics/btp348</a></p>
<h3>摘要</h3>
<p>Multiple sequence alignments are central to many areas of bioinformatics. It has been shown that the removal of poorly aligned regions from an alignment increases the quality of subsequent analyses. Such an alignment trimming phase is complicated in large-scale phylogenetic analyses that deal with thousands of alignments. Here, we present trimAl, a tool for automated alignment trimming, which is especially suited for large-scale phylogenetic analyses. trimAl can consider several parameters, alone or in multiple combinations, for selecting the most reliable positions in the alignment. These include the proportion of sequences with a gap, the level of amino acid similarity and, if several alignments for the same set of sequences are provided, the level of consistency across different alignments. Moreover, trimAl can automatically select the parameters to be used in each specific alignment so that the signal-to-noise ratio is optimized.<br>
多序列比对是生物信息学许多领域的核心。<br>
已经表明，从比对中去除对齐不良的区域提高了后续分析的质量。<br>
在处理数千个比对的大规模系统发育分析中，这样的比对修剪阶段是复杂的。<br>
在这里，我们提出了trimAl，自动对齐修剪，这是特别适合于大规模的系统发育分析的工具。<br>
trimAl可以单独或以多种组合考虑若干参数，用于选择比对中最可靠的位置。<br>
这些包括具有缺口的序列的比例、氨基酸相似性的水平，以及如果提供了同一组序列的几个比对，则不同比对之间的一致性水平。<br>
此外，trimAl可以自动选择在每个特定比对中使用的参数，从而优化信噪比。</p>
<p>可用性：trimAl是用C++编写的，可以移植到所有平台。trimAl可<a href="http://trimal.cgenomics.org" rel="nofollow">免费下载</a>，并可通过<a href="http://phylemon2.bioinfo.cipf.es/" rel="nofollow">Phylemon网络服务器</a>在线使用。补充材料可在 <a href="http://trimal.cgenomics.org/publications" rel="nofollow">http://trimal.cgenomics.org/publications</a> 上获得。</p>
<p>联系人：<a href="mailto:tgabaldon@crg.es">tgabaldon@crg.es</a></p>
<p>Multiple sequence alignments (MSA) are central to many areas of bioinformatics, including phylogenetics, homology modeling, database searches and motif finding. Recently, such MSA-based techniques have been incorporated in high-throughput pipelines such as genome annotation and phylogenomics analyses. In all these applications, the reliability and accuracy of the analyses depend critically on the quality of the underlying alignments. A plethora of computer programs and algorithms for MSA are currently available (Notredame, 2007), which implement different heuristics to find mathematically optimal solutions to the MSA problem. Accuracies of 80–90% have been reported for the best algorithms, but even the best scoring alignment algorithms may fail with certain protein families or at specific regions in the alignment. The situation worsens in large-scale analyses, where faster but less reliable algorithms and large numbers of automatically selected sequences are used. It is therefore generally assumed that trimming the alignment, so that poorly aligned regions are eliminated, increases the accuracy of the resulting MSA-based applications (Talavera and Castresana, 2007). Some programs such as G-blocks (Castresana, 2000) have been developed to assist in the MSA trimming phase by selecting blocks of conserved regions. They have become very popular and are extensively used, with good performance, in small-to-medium scale datasets, where several parameters can be tested manually (Talavera and Castresana, 2007). However, their use over larger datasets is hampered by the need for defining, prior to the analysis, the set of parameters that will be used for all sequence families. Here, we present trimAl, a tool for automated alignment trimming. Its speed and the possibility for automatically adjusting the parameters to improve the phylogenetic signal-to-noise ratio, makes trimAl especially suited for large-scale phylogenomic analyses, involving thousands of large alignments.<br>
多序列比对（MSA）是生物信息学许多领域的核心，包括遗传学、同源建模、数据库搜索和基序发现。最近，这种基于MSA的技术已被纳入高通量管道，如基因组注释和测序基因组学分析。在所有这些应用中，分析的可靠性和准确性关键取决于底层对齐的质量。目前有大量的MSA计算机程序和算法（Notredame，2007），它们实现了不同的算法来找到MSA问题的数学最优解。据报道，最好的算法的准确度为80-90%，但即使是最好的评分比对算法也可能在某些蛋白质家族或比对中的特定区域失败。这种情况在大规模分析中是不稳定的，在大规模分析中，使用更快但不太可靠的算法和大量自动选择的序列。 因此，通常认为，调整对齐，从而消除对齐不良的区域，可以提高基于MSA的应用程序的准确性（Talavera和Castresana，2007）。已经开发了一些程序，例如G-模块（Castresana，2000），以通过选择保守区域的模块来帮助MSA修剪阶段。它们已经变得非常流行，并在中小规模数据集中广泛使用，具有良好的性能，其中可以手动测试几个参数（Talavera和Castresana，2007）。然而，它们在较大数据集上的使用受到在分析之前定义将用于所有序列家族的参数集的需要的阻碍。在这里，我们提出了trimAl，一个自动对齐修剪工具。 它的速度和自动调整参数以提高系统发育信噪比的可能性，使得trimAl特别适合于大规模的基因组分析，涉及数千个大型比对。</p>
<p>trimAl has been developed in a GNU/Linux environment using C++ programming language and has been tested on various UNIX, Mac and Windows platforms. Moreover, we have developed a web server to run trimAl online (<a href="http://phylemon2.bioinfo.cipf.es/" rel="nofollow">http://phylemon2.bioinfo.cipf.es/</a>), which has been included in the Phylemon suite for phylogenetic and phylogenomic tools (Tarraga et al., 2007). The documentation, source files and additional information for trimAl are available through a wiki page (<a href="http://trimal.cgenomics.org/" rel="nofollow">http://trimal.cgenomics.org/</a>).<br>
trimAl是在GNU/Linux环境中使用C++编程语言开发的，并在各种UNIX、Mac和Windows平台上进行了测试。此外，我们已经开发了一个网络服务器来在线运行trimAl（http：//phylemon2.bioinfo.cipf.es/），其已经被包括在Phylemon套件中，用于系统发育和非系统发育基因组工具（Tarraga等人，2007年）。trimAl的文档、源文件和其他信息可通过wiki页面（<a href="http://trimal.cgenomics.org%EF%BC%89%E8%8E%B7%E5%BE%97%E3%80%82" rel="nofollow">http://trimal.cgenomics.org）获得。</a></p>
<p>trimAl reads and renders protein or nucleotide alignments in several standard formats. trimAl starts by reading all columns in an alignment and computes a score (Sx) for each of them. This score can be a gap score (Sg), a similarity score (Ss) or a consistency score (Sc). The score for each column can be computed based only on the information from that column or, if a window size of w is specified, it corresponds to the average value of w columns around the position considered.<br>
trimAl以几种标准格式读取并呈现蛋白质或核苷酸比对。trimAl通过阅读比对中的所有列开始，并为它们中的每一个计算得分（S x ）。该分数可以是间隙分数（S g ）、相似性分数（S s ）或一致性分数（S c ）。每个列的得分可以仅基于来自该列的信息来计算，或者如果指定窗口大小为w，则它对应于所考虑的位置周围的w列的平均值。</p>
<p>The gap score (Sg) for a column is the fraction of sequences without a gap in that position. The residue similarity score (Ss) consists of mean distance (MD) scores as described in Thompson et al. (2001) and <a href="https://academic.oup.com/bioinformatics/article/25/15/1972/213148?login=true#supplementary-data" rel="nofollow">Supplementary Material</a>. This score uses the MD between pairs of residues, as defined by a given scoring matrix. Finally, the consistency score (Sc) can only be computed when more than one alignment for the same set of sequences is provided. Details on how these scores are computed are provided in the <a href="https://academic.oup.com/bioinformatics/article/25/15/1972/213148?login=true#supplementary-data" rel="nofollow">Supplementary Material</a>. In brief, Sc measures the level of consistency of all the residue pairs found in a column as compared with the other alignments. The alignment with the highest consistency is chosen and then trimmed to remove the columns that are less conserved, according to Sc or other thresholds set by the user.<br>
一列差距分数（S g ）是在该位置没有缺口的序列的分数。残基相似性评分（S s ）由Thompson等人（2001）和补充材料中所述的平均距离（MD）评分组成。该评分使用残基对之间的MD，如给定评分矩阵所定义。最后，只有当提供了同一组序列的多于一个比对时，才能计算一致性得分（S c ）.关于如何计算这些分数的详细信息，请参见补充材料。简而言之，S c 测量了与其它比对相比，在一个柱中发现的所有残基对的一致性水平。根据S c 或用户设置的其他阈值，选择具有最高一致性的对齐，然后对其进行修整以移除较少保留的列。</p>
<p>Once all column scores have been computed trimAl can proceed in two ways. If both a score and a minimum conservation threshold are provided, trimAl renders a trimmed alignment in which only the columns with scores above the score threshold are included, as far as the number of selected columns is above a conservation threshold defined by the user. If this number is below the conservation threshold, trimAl will add more columns to the trimmed alignment in a decreasing order of scores until the conservation threshold is reached. The conservation threshold corresponds to the minimum percentage of columns, from the original alignment, which the user wants to include in the trimmed alignment.<br>
一旦计算了所有列得分，trimAl可以以两种方式进行。如果提供了分数和最小保守阈值两者，则trimAl呈现修剪的比对，其中仅包括分数高于分数阈值的列，只要所选列的数量高于由用户定义的保守阈值。如果该数量低于保守阈值，则trimAl将以分数的降序向修剪的比对添加更多列，直到达到保守阈值。保留阈值对应于原始路线中用户希望包含在修剪路线中的列的最小百分比。</p>
<p>Alternatively, if the automatic selection of parameters options is selected, trimAl will compute specific score thresholds depending on the inherent characteristics of each alignment. So far, trimAl incorporates three modes for the automated selection of parameters, gappyout, strict and strictplus, which are based on the different use of gap and similarity scores. Moreover, the option automated1 implements a heuristic to decide the most appropriate mode depending on the alignment characteristics. The heuristics to define such parameters have been designed based on the results of a benchmark. Details on the heuristics and the benchmark can be found in the online documentation of the program. In brief, the automatic selection of parameters approximate optimal cutoffs by plotting, internally, the cumulative graphs of gap and similarity scores of the columns in the alignment (see online documentation).<br>
或者，如果选择了参数选项的自动选择，则trimAl将根据每个比对的固有特性计算特定的分数阈值。到目前为止，trimAl集成了三种自动选择参数的模式，gappyout，strict和strictplus，它们基于gap和相似性分数的不同使用。此外，选项automated1实现了一种启发式方法，以根据对齐特性决定最合适的模式。基于基准测试的结果设计了定义这些参数的算法。有关算法和基准的详细信息可以在该程序的在线文档中找到。简而言之，参数的自动选择通过在内部绘制比对中列的间隙和相似性得分的累积图来近似最佳截止值（参见在线文档）。</p>
<p>We expanded, using ROSE simulations (Stoye et al., 1998) a benchmark set that has been used previously to test the improvement in phylogenetic performance after an alignment trimming phase (Talavera and Castresana, 2007). This dataset simulates several evolutionary scenarios varying in the number and length of the sequences, the topology of the underlying tree and the level of sequence divergence considered. We compared the results obtained from MUSCLE alignments before and after trimming with trimAl using automated selection of parameters. The accuracy of the resulting trees was measured by comparing them with the original trees used to generate the sequence sets, and measuring the Robinson Foulds distance (Robinson and Foulds, 1981). We observed an overall improvement of the phylogenetic accuracy after trimming. Using -automated1 option of trimAl, the trimmed alignment always produced Maximum Likelihood trees that were of equal (36%) or significantly better (64%) quality as compared with the tree derived from the complete alignment. For Neighbor Joining reconstruction the -strictplus option of trimAl worked best, improving the phylogenetic accuracy in 89% of the scenarios. In most scenarios (90%), trimAl outperformed Gblocks v0.91b with default parameters. Most importantly, the use of Gblocks default parameters diminished the accuracy of the subsequent tree reconstruction in half of the scenarios considered. In contrast, the use of trimAl automated methods rarely (1.5%) undermined the topological accuracy of the resulting phylogenetic tree (see <a href="https://academic.oup.com/bioinformatics/article/25/15/1972/213148?login=true#supplementary-data" rel="nofollow">Supplementary Material</a> for more details).<br>
我们扩展，使用ROSE模拟（Stoye等人，1998），这是一个基准集，以前曾用于测试比对修剪阶段后系统发育性能的改善（Talavera和Castresana，2007）。该数据集模拟了几种进化场景，这些场景在序列的数量和长度、底层树的拓扑结构和所考虑的序列分歧水平方面有所不同。我们使用自动选择的参数比较了在用trimAl修剪之前和之后从肌肉比对获得的结果。通过将所得树与用于生成序列集的原始树进行比较，并测量罗宾逊福尔德距离（罗宾逊和福尔德，1981），来测量所得树的准确性。我们观察到修剪后的系统发育准确性的整体改善。 使用trimAl的-automated 1选项，修剪的比对总是产生最大似然树，与从完全比对中得到的树相比，其质量相等（36%）或显著更好（64%）。对于相邻连接重建，trimAl的-strictplus选项工作得最好，在89%的场景中提高了系统发育的准确性。在大多数情况下（90%），trimAl优于默认参数的Gblocks v0.91b。最重要的是，在所考虑的一半场景中，使用Gblocks默认参数降低了后续树重建的准确性。相比之下，使用trimAl自动化方法很少（1.5%）破坏所得到的系统发育树的拓扑准确性（更多细节见补充材料）。</p>
<p>To test the applicability of trimAl on real datasets as well as its suitability for large-scale phylogenetic datasets, we ran trimAl on the complete set of MUSCLE alignments generated for the Human Phylome project (Huerta-Cepas et al., 2007). This includes a total of 31 182 alignments, containing, on average, 67 sequences of 1472 positions of length. Trimming these alignments using the -gappyout and automated1 options used 5 min 45 s and 125 min, 2 s, respectively, on a computer with an Intel QuadCore XEON E5410 processors and 8 GB of RAM.<br>
为了测试trimAl在真实的数据集上的适用性以及其对大规模系统发育数据集的适用性，我们在为人类Phylome项目生成的完整的肌肉比对集上运行trimAl（Huerta—Cepas等人，2007年）。这包括总共31182个比对，平均包含67个序列，长度为1472个位置。在配备英特尔四核至强E5410处理器和8 GB RAM的计算机上，使用—gappyout和automated1选项修剪这些比对分别需要5 min 45 s和125 min 2 s。</p>
<p>trimAl has been used previously in a pipeline to reconstruct complete collections of gene trees. In this case, the parameter sets used were a minimum conservation threshold of 60% and a gap threshold of 90% (-cons 60 -gt 0.9). Complete and trimmed alignments used to generate the phylomes included in PhylomeDB (Huerta-Cepas et al., 2008) can be viewed through this database.<br>
trimAl先前已经在管道中用于重建基因树的完整集合。在这种情况下，使用的参数集是60%的最小保守阈值和90%的间隙阈值（-cons 60 -gt 0.9）。用于生成PhylomeDB中包括的系统发育组的完整和修剪的比对（Huerta-Cepas等人，2008年，可以通过这个数据库。</p></div>
<div style="font-size:small;margin-top:8px;float:right;">❤️ 转载文章请注明出处，谢谢！❤️</div>

<button class="btn btn-block" type="button" onclick="openComments()" id="cmButton">评论</button>
<div class="comments" id="comments"></div>

</div>
    <div id="footer"><span id="hitokoto_text">:D</span> —— <span id="hitokoto_from_who">:D</span>「<span id="hitokoto_from">:D</span>」<br><span id="busuanzi">❤️感谢第<span></span>位小伙伴的<span></span>次访问本站。</span>
<p>
<span id="runday"></span>Powered by <a href="https://meekdai.com/Gmeek.html" target="_blank">Gmeek</a>
</p>

<script>
if("03/30/2019"!=""){
    var now=new Date();
    var startSite=new Date("03/30/2019");
    var diff=now.getTime()-startSite.getTime();
    var diffDay=Math.floor(diff/(1000*60*60*24));
    document.getElementById("year").innerHTML=now.getFullYear();
    if(""!=""){document.getElementById("runday").innerHTML=" • "+"网站运行"+diffDay+"天"+" • ";}
    else{document.getElementById("runday").innerHTML="网站运行"+diffDay+"天"+" • ";}
}
</script>

<script>document.getElementById('busuanzi').id='busuanzi_container_site_uv';busuanzi=document.getElementById('busuanzi_container_site_uv');busuanzi.style.display='none';busuanzi.childNodes[1].id='busuanzi_value_site_uv';busuanzi.childNodes[3].id='busuanzi_value_site_pv';</script><script async src='//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js'></script>

<script>
	fetch('https://v1.hitokoto.cn')
	  .then(response => response.json())
	  .then(data => {
		const hitokoto_text = document.querySelector('#hitokoto_text')
		hitokoto_text.innerText = data.hitokoto
		const hitokoto_from = document.querySelector('#hitokoto_from')
		hitokoto_from.innerText = data.from
		const hitokoto_from_who = document.querySelector('#hitokoto_from_who')
		hitokoto_from_who.innerText = data.from_who
	  })
	  .catch(console.error)
  </script>

<style>#busuanzi_value_site_uv{color:red}#busuanzi_value_site_pv{color:red}</style></div>
</body>
<script>
var IconList={'sun': 'M8 10.5a2.5 2.5 0 100-5 2.5 2.5 0 000 5zM8 12a4 4 0 100-8 4 4 0 000 8zM8 0a.75.75 0 01.75.75v1.5a.75.75 0 01-1.5 0V.75A.75.75 0 018 0zm0 13a.75.75 0 01.75.75v1.5a.75.75 0 01-1.5 0v-1.5A.75.75 0 018 13zM2.343 2.343a.75.75 0 011.061 0l1.06 1.061a.75.75 0 01-1.06 1.06l-1.06-1.06a.75.75 0 010-1.06zm9.193 9.193a.75.75 0 011.06 0l1.061 1.06a.75.75 0 01-1.06 1.061l-1.061-1.06a.75.75 0 010-1.061zM16 8a.75.75 0 01-.75.75h-1.5a.75.75 0 010-1.5h1.5A.75.75 0 0116 8zM3 8a.75.75 0 01-.75.75H.75a.75.75 0 010-1.5h1.5A.75.75 0 013 8zm10.657-5.657a.75.75 0 010 1.061l-1.061 1.06a.75.75 0 11-1.06-1.06l1.06-1.06a.75.75 0 011.06 0zm-9.193 9.193a.75.75 0 010 1.06l-1.06 1.061a.75.75 0 11-1.061-1.06l1.06-1.061a.75.75 0 011.061 0z', 'moon': 'M9.598 1.591a.75.75 0 01.785-.175 7 7 0 11-8.967 8.967.75.75 0 01.961-.96 5.5 5.5 0 007.046-7.046.75.75 0 01.175-.786zm1.616 1.945a7 7 0 01-7.678 7.678 5.5 5.5 0 107.678-7.678z', 'sync': 'M1.705 8.005a.75.75 0 0 1 .834.656 5.5 5.5 0 0 0 9.592 2.97l-1.204-1.204a.25.25 0 0 1 .177-.427h3.646a.25.25 0 0 1 .25.25v3.646a.25.25 0 0 1-.427.177l-1.38-1.38A7.002 7.002 0 0 1 1.05 8.84a.75.75 0 0 1 .656-.834ZM8 2.5a5.487 5.487 0 0 0-4.131 1.869l1.204 1.204A.25.25 0 0 1 4.896 6H1.25A.25.25 0 0 1 1 5.75V2.104a.25.25 0 0 1 .427-.177l1.38 1.38A7.002 7.002 0 0 1 14.95 7.16a.75.75 0 0 1-1.49.178A5.5 5.5 0 0 0 8 2.5Z', 'home': 'M6.906.664a1.749 1.749 0 0 1 2.187 0l5.25 4.2c.415.332.657.835.657 1.367v7.019A1.75 1.75 0 0 1 13.25 15h-3.5a.75.75 0 0 1-.75-.75V9H7v5.25a.75.75 0 0 1-.75.75h-3.5A1.75 1.75 0 0 1 1 13.25V6.23c0-.531.242-1.034.657-1.366l5.25-4.2Zm1.25 1.171a.25.25 0 0 0-.312 0l-5.25 4.2a.25.25 0 0 0-.094.196v7.019c0 .138.112.25.25.25H5.5V8.25a.75.75 0 0 1 .75-.75h3.5a.75.75 0 0 1 .75.75v5.25h2.75a.25.25 0 0 0 .25-.25V6.23a.25.25 0 0 0-.094-.195Z', 'github': 'M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z', 'copy': 'M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z', 'check': 'M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z'};
var utterancesLoad=0;

let themeSettings={
    "dark": ["dark","moon","#00f0ff","dark-blue"],
    "light": ["light","sun","#ff5000","github-light"],
    "auto": ["auto","sync","","preferred-color-scheme"]
};
function changeTheme(mode, icon, color, utheme){
    document.documentElement.setAttribute("data-color-mode",mode);
    document.getElementById("themeSwitch").setAttribute("d",value=IconList[icon]);
    document.getElementById("themeSwitch").parentNode.style.color=color;
    if(utterancesLoad==1){utterancesTheme(utheme);}
}
function modeSwitch(){
    let currentMode=document.documentElement.getAttribute('data-color-mode');
    let newMode = currentMode === "light" ? "dark" : currentMode === "dark" ? "auto" : "light";
    localStorage.setItem("meek_theme", newMode);
    if(themeSettings[newMode]){
        changeTheme(...themeSettings[newMode]);
    }
}
function utterancesTheme(theme){
    const message={type:'set-theme',theme: theme};
    const iframe=document.getElementsByClassName('utterances-frame')[0];
    iframe.contentWindow.postMessage(message,'https://utteranc.es');
}
if(themeSettings[theme]){changeTheme(...themeSettings[theme]);}
console.log("\n %c Gmeek main https://github.com/Meekdai/Gmeek \n","padding:5px 0;background:#02d81d;color:#fff");
</script>

<script>
document.getElementById("pathHome").setAttribute("d",IconList["home"]);
document.getElementById("pathIssue").setAttribute("d",IconList["github"]);



function openComments(){
    cm=document.getElementById("comments");
    cmButton=document.getElementById("cmButton");
    cmButton.disabled=true;
    cmButton.innerHTML="loading";
    span=document.createElement("span");
    span.setAttribute("class","AnimatedEllipsis");
    cmButton.appendChild(span);

    script=document.createElement("script");
    script.setAttribute("src","https://utteranc.es/client.js");
    script.setAttribute("repo","seeu100/blog");
    script.setAttribute("issue-term","title");
    
    if(localStorage.getItem("meek_theme")=="dark"){script.setAttribute("theme","dark-blue");}
    else if(localStorage.getItem("meek_theme")=="light") {script.setAttribute("theme","github-light");}
    else{script.setAttribute("theme","preferred-color-scheme");}
    
    script.setAttribute("crossorigin","anonymous");
    script.setAttribute("async","");
    cm.appendChild(script);

    int=self.setInterval("iFrameLoading()",200);
}

function iFrameLoading(){
    var utterances=document.getElementsByClassName('utterances');
    if(utterances.length==1){
        if(utterances[0].style.height!=""){
            utterancesLoad=1;
            int=window.clearInterval(int);
            document.getElementById("cmButton").style.display="none";
            console.log("utterances Load OK");
        }
    }
}

document.addEventListener('DOMContentLoaded', () => {
    const createClipboardHTML = (codeContent, additionalClasses = '') => `
        <pre class="notranslate"><code class="notranslate">${codeContent}</code></pre>
        <div class="clipboard-container position-absolute right-0 top-0 ${additionalClasses}">
            <clipboard-copy class="ClipboardButton btn m-2 p-0" role="button" style="display: inherit;">
                <svg height="16" width="16" class="octicon octicon-copy m-2"><path d="${IconList["copy"]}"></path></svg>
                <svg height="16" width="16" class="octicon octicon-check color-fg-success m-2 d-none"><path d="${IconList["check"]}"></path></svg>
            </clipboard-copy>
            <div class="copy-feedback">Copied!</div>
        </div>
    `;

    const handleCodeElements = (selector = '') => {
        document.querySelectorAll(selector).forEach(codeElement => {
            const codeContent = codeElement.innerHTML;
            const newStructure = document.createElement('div');
            newStructure.className = 'snippet-clipboard-content position-relative overflow-auto';
            newStructure.innerHTML = createClipboardHTML(codeContent);

            const parentElement = codeElement.parentElement;
            if (selector.includes('highlight')) {
                parentElement.insertBefore(newStructure, codeElement.nextSibling);
                parentElement.removeChild(codeElement);
            } else {
                parentElement.parentElement.replaceChild(newStructure, parentElement);
            }
        });
    };

    handleCodeElements('pre.notranslate > code.notranslate');
    handleCodeElements('div.highlight > pre.notranslate');

    let currentFeedback = null;
    document.querySelectorAll('clipboard-copy').forEach(copyButton => {
        copyButton.addEventListener('click', () => {
            const codeContent = copyButton.closest('.snippet-clipboard-content').innerText;
            const tempTextArea = document.createElement('textarea');
            tempTextArea.value = codeContent;
            document.body.appendChild(tempTextArea);
            tempTextArea.select();
            document.execCommand('copy');
            document.body.removeChild(tempTextArea);

            const copyIcon = copyButton.querySelector('.octicon-copy');
            const checkIcon = copyButton.querySelector('.octicon-check');
            const copyFeedback = copyButton.nextElementSibling;

            if (currentFeedback && currentFeedback !== copyFeedback) {currentFeedback.style.display = 'none';}
            currentFeedback = copyFeedback;

            copyIcon.classList.add('d-none');
            checkIcon.classList.remove('d-none');
            copyFeedback.style.display = 'block';
            copyButton.style.borderColor = 'var(--color-success-fg)';

            setTimeout(() => {
                copyIcon.classList.remove('d-none');
                checkIcon.classList.add('d-none');
                copyFeedback.style.display = 'none';
                copyButton.style.borderColor = '';
            }, 2000);
        });
    });
});

</script>


</html>
